Tiny MCE plugin overrides custom settings

DerinMavi opened the pull request:
Pull Request for Issue # .

### Summary of Changes
Issue reported here: #9812, #8591, #5820, #11851 and #8106
Some of the issues were closed because of the pull requests that came but later it was found out either they do not work or they are wrong(they modify default values).

This change fixes the problem while keeping the default values.

### Testing Instructions
Please test all 3 use cases:

1) Use the following code to display a tinymce editor:
`
$editor = JEditor::getInstance('tinymce');
echo $editor->display('description', "", '250', '50', '60', '5', false);`

2) Add a new article from backend of the site, try to edit this xml code https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_content/models/forms/article.xml#L27 , add width and height attributes (see https://docs.joomla.org/Editor_form_field_type for document about editor field type). For example, change it to;
`<field name="articletext" type="editor"
			label="COM_CONTENT_FIELD_ARTICLETEXT_LABEL" description="COM_CONTENT_FIELD_ARTICLETEXT_DESC"
			filter="JComponentHelper::filterText" buttons="true"  height="50" width="250" />`

3) In the joomla backend go to extensions -> plugins -> tinymce editor -> advanced settings and modify width & height

### Expected result
Tinymce editor has height of 50pixels and width of 250pixels
### Actual result
Tinymce editor has height of 550pixels
### Documentation Changes Required

DerinMavi committed:
Fixed: Tiny MCE plugin overrides custom settings

Issue reported here: #9812, #8591, #5820, #11851 and #8106
Some of the issues were closed when the pull requests came but later it was found out PR does not work/modify default values.

This fixes the problem while keeping the default values.

DerinMavi committed:
Update tinymce.php

DerinMavi commented:
I don't understand why the Travis CI build failed, any ideas?

infograf768 commented:
Travis says that your formatting is wrong:
FILE: ...home/travis/build/joomla/joomla-cms/plugins/editors/tinymce/tinymce.php

--------------------------------------------------------------------------------

FOUND 2 ERROR(S) AFFECTING 2 LINE(S)

--------------------------------------------------------------------------------

 363 | ERROR | Expected "if (...)\n...{...}\n...else\n"; found

     |       | "if(...)\n...{...}\n...else "

 371 | ERROR | Expected "if (...)\n...{...}\n...else\n"; found

     |       | "if(...)\n...{...}\n...else\n"

i.e. Instead of

![screen shot 2017-02-12 at 10 09 47](https://cloud.githubusercontent.com/assets/869724/22860911/70370d3e-f10b-11e6-9c36-9c0251216037.png)
You should have

![screen shot 2017-02-12 at 10 11 22](https://cloud.githubusercontent.com/assets/869724/22860921/acf83590-f10b-11e6-9178-62b264373c58.png)

(Note: Did not test patch itself)

DerinMavi committed:
fixed indentation

DerinMavi committed:
Update tinymce.php

DerinMavi committed:
removed tabs from empty lines

DerinMavi committed:
moved opening brace to new line

DerinMavi commented:
Thanks! Fixed the whitespace issues.

infograf768 commented:
This works, except that the test should be:
```php
$editor = JEditor::getInstance('tinymce'); 
echo $editor->display('description', "", '250', '50', '60', '5', false);

```
$width comes before $height
See https://github.com/joomla/joomla-cms/blob/staging/libraries/cms/editor/editor.php#L289

infograf768 commented:
I have tested this item :white_check_mark: successfully on 984c0c7199fd38cd5dcfb17bc7f54854d7a42db2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14033">issues.joomla.org/tracker/joomla-cms/14033</a>.</sub>

infograf768 commented:
@Fedik @dgt41 
Please test. I guess this can go in.

AlexRed commented:
but the advanced params still not working
![http://alexred.com/joomla/2017/tinyMCE-advanced.png](http://alexred.com/joomla/2017/tinyMCE-advanced.png "http://alexred.com/joomla/2017/tinyMCE-advanced.png")<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14033">issues.joomla.org/tracker/joomla-cms/14033</a>.</sub>

DerinMavi commented:
@AlexRed If those settings did not work before, I guess it is a separate issue. This fix is only for 3rd party components that programmatically display the editor.

AlexRed commented:
Before your patch the "HTML Height" parameter work, after not work  :(

infograf768 commented:
I confirm @Alexred findings and changed my test to negative.

infograf768 commented:
The issue is that $height has always a value.
To test, just add a  var_dump before checking height
```
		var_dump($height);
		die;
		if (isset($height))
		{
			$html_height = $height;
		}
		else
		{
			$html_height = $this->params->get('html_height', '550');
		}
```
$height will default to 500px

infograf768 commented:
so, tho get this to work (for height only), we would have to use
```php
                if ($height != 500)
		{
			$html_height = $height;
		}
		else
		{
			$html_height = $this->params->get('html_height', '550');
		}
```

joomdonation commented:
isset is the wrong command and could not be used here. The code should be changed to:

```php
if ($height)
{
	$html_height = $height;
}
else
{
	$html_height = $this->params->get('html_height', '550');
}

if ($width)
{
	$html_width  = $width;
}
else
{
	$html_width  = $this->params->get('html_width', '');
}
```

With this change, custom width (and height) if defined will be used instead of the width and height defined in plugins parameter

The easiest way is try to add a new article from backend of the site, try to edit this xml code https://github.com/joomla/joomla-cms/blob/staging/administrator/components/com_content/models/forms/article.xml#L27 , add width and height attributes (see https://docs.joomla.org/Editor_form_field_type for document about editor field type). For example, change it to;

```xml
<field name="articletext" type="editor"
			label="COM_CONTENT_FIELD_ARTICLETEXT_LABEL" description="COM_CONTENT_FIELD_ARTICLETEXT_DESC"
			filter="JComponentHelper::filterText" buttons="true"  height="300" width="300" />
```

1. Before the change, the custom with and height is not used. The width and height parameters defined in editor plugin is used

2. After changed, the custom height and width passed is used (note, size of editor changed)

3. Remove width, height attribute from XML code above, the editor again uses the size defined in the plugin parameter

Please note that in case the form has multiple editors, the width and height defined in the first displayed editors will be used. If you define difference size for different editor, it won't have any affect, the size defined in the first displayed editor always win.

infograf768 commented:
@joomdonation
As I wrote above, we always have a default height of 500px, therefore it kills any height defined in the plugin params.

joomdonation commented:
@infograf768 How did you test it? Please try to change the XML file of add article form as I suggested, you will see that it works

infograf768 commented:
Do not change anything. Do not patch or if you patch, do the var_dump as I have shown above.
```
var_dump($height);
		die;
```
We do have a default 500px coming from I don't know where (maybe from tiny js files)

joomdonation commented:
Ah, see your point. See https://github.com/joomla/joomla-cms/blob/staging/libraries/cms/form/field/editor.php#L204-L205

If we don't define width and height, default values are passed there instead of using editor parameters. No sure what's the reasons

infograf768 commented:
> If we don't define width and height, default values are passed there instead of using editor parameters. No sure what's the reasons

Bingo. it was there

joomdonation commented:
In this case, I think the best fix we could do is changing the code to:

```php
if ($height && $height != '500px')
{
	$html_height = $height;
}
else
{
	$html_height = $this->params->get('html_height', '550');
}

if ($width && $width != '100%')
{
	$html_width  = $width;
}
else
{
	$html_width  = $this->params->get('html_width', '');
}
```

infograf768 commented:
Not sure we need
`if ($height && $height != '500px')`

as `$height` always exists.

Also, no need for `500px`. My test show that `500` is fine`

joomdonation commented:
> Not sure we need
> if ($height && $height != '500px')

=> Yes, we need it to give more flexible to developer. For example, when you call

```php
echo $editor->display('description', "", '', '', '60', '5', false);
```
The editor should be rendered using width and height from plugins parameter

> Also, no need for 500px. My test show that 500 is fine`

When no height is being passed, $height value is set to 500px (it is a string). I guess your code works because php doing magic comparison. For example the code below display true when the two value actually not equals

var_dump(500 == '500px');

infograf768 commented:
> The editor should be rendered using width and height from plugins parameter

Are you sure it does not pick the default from
```
			$this->height      = $this->element['height'] ? (string) $this->element['height'] : '500';
			$this->width       = $this->element['width'] ? (string) $this->element['width'] : '100%';
```

joomdonation commented:
It picks the default value if we are using editor form field type. It doesn't have default value when you call the $editor->display directly (in the code you gave for testing)

DerinMavi commented:
I will try to fix it today.

DerinMavi committed:
Update tinymce.php

DerinMavi committed:
Update codemirror.php

default values

DerinMavi committed:
added default values

DerinMavi committed:
default width/height values are set in onDisplay

DerinMavi commented:
@joomdonation : The problem with your suggestion is that if you set height: '500px' or width:'100%' in the article.xml , your settings will be ignored and they will be overridden by tinymce advanced settings parameters.

The root cause of the problem is that for tinymce there are 2 places where default width/height is provided. Once onDisplay method inside tinymce.php is called it is impossible to know whether the widht/height are default values or user set values.

For this reason, I removed the default values from setup() method in editor.php.

For all editors, default values are set inside onDisplay() methods.

Please @infograf768 and @AlexRed test the new fix.

joomdonation commented:
@DerinMavi The reason I suggested it because I was worry that if we remove default values in setup() method (like how you did in recent commit), there might be backward compatible issue with third party editor plugins. 

Right now, if no width and height provided in the XML form field, third party editor plugins might expect that default width and height provided by setup method and they might not deal with empty value (like you modified for the editor plugin shipped by Joomla). That's why I mentioned in my comment that _best fix we could do_

DerinMavi committed:
reverting changes

DerinMavi committed:
reverting changes

DerinMavi committed:
new fix

DerinMavi committed:
revertint change

DerinMavi committed:
removed whitespace

DerinMavi committed:
Update tinymce.php

DerinMavi commented:
@joomdonation You are right, I checked couple of 3rd party editors and saw that they'd break if the defaults are not set. I finally settled with your suggested solution. I changed `!=` with `!==`(so that if someone passes `500` (integer) from the code it'd work as expected. Also changed the default value of `550px` to `500px` so that it is consistent with previous default value.

DerinMavi committed:
500px to 500

 so that it is consistent with previous code

joomdonation commented on plugins/editors/tinymce/tinymce.php:
```diff
@@ -360,12 +360,24 @@ public function onDisplay($name, $content, $width, $height, $col, $row, $buttons
-		$html_height = $this->params->get('html_height', '550');
-		$html_width  = $this->params->get('html_width', '');
+		// Compare the value with what is set in JFormFieldEditor::setup()
+		if ($height && $height !== '500')
```
This need to be changed to if ($height && $height !== '500px') . The reason is because in the previous section of the code https://github.com/joomla/joomla-cms/blob/staging/plugins/editors/tinymce/tinymce.php#L149-L152 , we append px into $width if a numeric value is passed.

joomdonation commented on plugins/editors/tinymce/tinymce.php:
```diff
@@ -360,12 +360,24 @@ public function onDisplay($name, $content, $width, $height, $col, $row, $buttons
+			$html_height = $this->params->get('html_height', '500');
+		}
 
-		if ($html_width == 750)
```
I don't know the reason why there was code to compare with 750 here, but I think it would be safe to leave it as how it was (don't remove that block of code)

DerinMavi replied:
That if check doesn't make any sense, no way I'm putting that code back. Adding random hard-coded if checks are not correct way to fix anything.

joomdonation replied:
Yes, it doesn't make sense, but there must be a reason the code is there. I am just worry about an accident backward compatible issue. Not a big thing, thought. I am going to test this change and report the result back

joomdonation replied:
Found the reason why the code is there.  Joomla set value of **HTML Width** parameter to **750** on default installation (see this screenshot https://drive.google.com/file/d/0B2RoJ_24efMQa3o1NE00TWwxVzQ/view?usp=drivesdk). So I think the reason of the code is to reset it to blank in case no custom value is entered and allow TinyMCE script to handle it.

With that said, I would suggest you to put the block of code back. When it is done, I will mark the test success.

DerinMavi committed:
fixed the width

500 => 500px

infograf768 commented:
We had a reason for `if ($html_width == 750)`, please add back.

Fedik commented on plugins/editors/tinymce/tinymce.php:
```diff
@@ -360,12 +360,24 @@ public function onDisplay($name, $content, $width, $height, $col, $row, $buttons
+		}
+		else
+		{
+			$html_width = $this->params->get('html_width', '100%');
```
should be 750 here, it is default value of the plugin, 
and the width/height for TinyMCE script should be in pixels https://www.tinymce.com/docs/configure/editor-appearance/#width

DerinMavi replied:
@Fedik Makes sense, fixed it.

DerinMavi committed:
tinymce width should be an integer

Fedik commented on plugins/editors/tinymce/tinymce.php:
```diff
@@ -360,12 +360,24 @@ public function onDisplay($name, $content, $width, $height, $col, $row, $buttons
+		}
+		else
+		{
+			$html_width = $this->params->get('html_width', '750');
```
so, now, do you see why was need? :neckbeard: 
```php
if ($html_width == 750)
{
   $html_width = '';
}
```

DerinMavi commented:
@joomdonation

N6REJ commented:
TinyMCE sizing does not work with form fields either, and I don't mean the new ones..
						name="brandText"
						class=""
						type="editor"
						default=""
						cols="2"
						rows="3"
						width="300px"
						height="300px"

schnuti commented:
@N6REJ 
width="300px"
height="300px"
Numeric! remove px and it should be ok
px is added later

rdeutz committed:
Merge branch 'staging' into staging

studio42 commented:
I'm on programatically change the editor height, this is not fixed in Joomla 3.8.6 !
Now more then1 year for this ?
Come on guys, it's not a big deal to check for values and use right ones.

studio42 commented:
In /plugins/editors/tinymce/tinymce.php line 408
```
		if($height) $html_height = $height;
		else $html_height = $this->params->get('html_height', '550');
```
Set the height correctly.

Schmidie64 commented:
I have tested this item :red_circle: unsuccessfully on d981c6b417cfce126e3c08b5e85fb9194fd21344<br /><br />I followed the instructions step by step but the patch didn't work well. Before i applied the patch i could change the editor size in the web config well but not in the xml file. After i applied the patch, i couldn't change anything, not in the web config neither nor in the xml File. So i thing the patch is not completely correct.

@icampus<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14033">issues.joomla.org/tracker/joomla-cms/14033</a>.</sub>

Quy commented:
Replaced by PR #26755<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/14033">issues.joomla.org/tracker/joomla-cms/14033</a>.</sub>
